src/connector/src/sink/mysql.rs

```diff
@@ -103,92 +125,94 @@ impl TryFrom<Datum> for MySQLValue {
-                            "UPDATE {} SET {} WHERE {};",
-                            self.table(),
-                            join(conditions(values(idx2)?), ","),
-                            join(conditions(values(idx)?), " AND ")
-                        )
-                    } else {
-                        panic!("UpdateDelete should always be followed by an UpdateInsert!")
-                    }
-                }
-                _ => panic!("UpdateInsert should always follow an UpdateDelete!"),
-            };
-            transaction.exec_drop(stmt, Params::Empty).await?;
-        }
-
-        // Commit and drop the connection.
-        transaction.commit().await?;
-        drop(conn);
+        self.chunk_cache.push((chunk, schema.clone()));
         Ok(())
     }
 
     async fn begin_epoch(&mut self, _epoch: u64) -> Result<()> {
-        todo!()
+        Ok(())
     }
 
     async fn commit(&mut self) -> Result<()> {
-        todo!()
+        let mut txn = self.conn.start_transaction(TxOpts::default()).await?;
+        for (chunk, schema) in &self.chunk_cache {
+            write_to_mysql(&mut txn, chunk, schema, &self.cfg).await?;
+        }
+        txn.commit().await?;
+
+        self.chunk_cache.clear();
+        Ok(())
     }
 
     async fn abort(&mut self) -> Result<()> {
-        todo!()
+        Ok(())
+    }
+}
+
+async fn write_to_mysql<'a>(
+    txn: &mut Transaction<'a>,
+    chunk: &StreamChunk,
+    schema: &Schema,
+    config: &MySQLConfig,
+) -> Result<()> {
+    // Closure that takes an idx to create a vector of MySQLValues from a StreamChunk 'row'.
+    let values = |idx| -> Result<Vec<MySQLValue>> {
+        chunk
+            .columns()
+            .iter()
+            .map(|x| MySQLValue::try_from(x.array_ref().datum_at(idx)))
+            .collect_vec()
+            .into_iter()
+            .collect()
+    };
+
+    // Closure that builds a String containing WHERE conditions, i.e. 'v1=1 AND v2=2'.
+    // Perhaps better to replace this functionality with a new Sink trait method.
+    let conditions = |values: Vec<MySQLValue>| {
+        schema
+            .names()
+            .iter()
+            .zip_eq(values.iter())
+            .map(|(c, v)| format!("{}={}", c, v))
+            .collect::<Vec<String>>()
+    };
+
+    let mut iter = chunk.ops().iter().enumerate();
+    while let Some((idx, op)) = iter.next() {
+        // Get SQL statement
+        let stmt = match *op {
+            Insert => format!(
+                "INSERT INTO {} VALUES ({});",
+                &config.table,
+                join(values(idx)?, ",")
+            ),
+            Delete => format!(
+                "DELETE FROM {} WHERE ({});",
+                &config.table,
+                join(conditions(values(idx)?), " AND ")
+            ),
+            UpdateDelete => {
+                if let Some((idx2, UpdateInsert)) = iter.next() {
+                    format!(
+                        "UPDATE {} SET {} WHERE {};",
```

Comment:
It looks like a bad idea to directly concat SQL into a string. It is prune to SQL injection attacks.

Reply:
e.g. it is possible that `insert xx values ("; DROP database");` in RisingWave will be sink to MySQL and eventually drop the whole database.

Reply:
I have the same concern. We may offer a DBC solution rather than concat SQL

Reply:
> e.g. it is possible that insert xx values ("; DROP database"); in RisingWave will be sink to MySQL and eventually drop the whole database.

I think `mysql_async` have done enough, for string, it will be converted to bytes and will not be treated as a separate SQL.
![image](https://user-images.githubusercontent.com/14017336/181059841-5cb2ef88-641e-4b73-8da7-33cfa8bb4ea6.png)

so I think we can merge this pr now

Reply:
I see. 

https://github.com/blackbeam/rust_mysql_common/blob/974c13b5275f9fcdff01a286ca4678afb12a9375/src/value/mod.rs#L286-L295

But it still sounds like a bad practice to concat SQL by hand. It's inefficient than using the wire protocol to compose the values. I think we will eventually go for the way like:

```
    r"INSERT INTO payment (customer_id, amount, account_name)
      VALUES (?, ?, ?)"
        .with(vec![mysql_common::Value(x), mysql_common::Value(y), mysql_common::Value(z)])
        .batch(&mut conn)
        .await?;
```

LGTM for me.